File: trunk/download/stream/src/test/java/org/simpleframework/xml/load/MapNullTest.java
package org.simpleframework.xml.load;

import java.io.StringWriter;
import java.math.BigDecimal;
import java.util.HashMap;
import java.util.Map;

import org.simpleframework.xml.Element;
import org.simpleframework.xml.ElementMap;
import org.simpleframework.xml.Root;
import org.simpleframework.xml.Serializer;
import org.simpleframework.xml.ValidationTestCase;

public class MapNullTest extends ValidationTestCase {
   
   private static final String ENTRY_MAP =     
   "<entryMap>\n"+
   "   <map>\n"+   
   "      <entry key='a'>" +
   "         <mapEntry>\n" +  
   "            <name>a</name>\n"+
   "            <value>example 1</value>\n"+
   "         </mapEntry>" + 
   "      </entry>" +
   "      <entry key='b'>" +
   "         <mapEntry>\n" +  
   "            <name>b</name>\n"+
   "            <value>example 2</value>\n"+
   "         </mapEntry>" + 
   "      </entry>" +
   "      <entry key='c'>" +
   "         <mapEntry>\n" +  
   "            <name>c</name>\n"+
   "            <value>example 3</value>\n"+
   "         </mapEntry>" + 
   "      </entry>" +
   "      <entry key='d'>" +
   "         <mapEntry>\n" +  
   "            <name>d</name>\n"+
   "            <value>example 4</value>\n"+
   "         </mapEntry>" + 
   "      </entry>" +
   "   </map>\n"+
   "</entryMap>";  
   
   private static final String STRING_MAP =     
   "<stringMap>\n"+
   "   <map>\n"+   
   "      <entry letter='a'>example 1</entry>\n" +
   "      <entry letter='b'>example 2</entry>\n" +
   "      <entry letter='c'>example 3</entry>\n" +
   "      <entry letter='d'>example 4</entry>\n" +   
   "   </map>\n"+
   "</stringMap>";
   
   private static final String COMPLEX_MAP =     
   "<complexMap>\n"+
   "   <map>\n"+   
   "      <entry>" +
   "         <compositeKey>\n" +
   "            <name>name 1</name>\n" +
   "            <address>address 1</address>\n" +
   "         </compositeKey>\n" +
   "         <mapEntry>\n" +  
   "            <name>a</name>\n"+
   "            <value>example 1</value>\n"+
   "         </mapEntry>" + 
   "      </entry>" +
   "      <entry>" +
   "         <compositeKey>\n" +
   "            <name>name 2</name>\n" +
   "            <address>address 2</address>\n" +
   "         </compositeKey>\n" +
   "         <mapEntry>\n" +  
   "            <name>b</name>\n"+
   "            <value>example 2</value>\n"+
   "         </mapEntry>" + 
   "      </entry>" +
   "      <entry>" +
   "         <compositeKey>\n" +
   "            <name>name 3</name>\n" +
   "            <address>address 3</address>\n" +
   "         </compositeKey>\n" +
   "         <mapEntry>\n" +  
   "            <name>c</name>\n"+
   "            <value>example 3</value>\n"+
   "         </mapEntry>" + 
   "      </entry>" +
   "      <entry>" +
   "         <compositeKey>\n" +
   "            <name>name 4</name>\n" +
   "            <address>address 4</address>\n" +
   "         </compositeKey>\n" +
   "         <mapEntry>\n" +  
   "            <name>d</name>\n"+
   "            <value>example 4</value>\n"+
   "         </mapEntry>" + 
   "      </entry>" +
   "   </map>\n"+
   "</complexMap>"; 
   
   private static final String PRIMITIVE_MAP =      
   "<primitiveMap>\n"+
   "   <table>\n"+   
   "      <entry>\n" +
   "         <string>one</string>\n" +
   "         <bigDecimal>1.0</bigDecimal>\n" +
   "      </entry>\n"+
   "      <entry>" +
   "         <string>two</string>\n" +
   "         <bigDecimal>2.0</bigDecimal>\n" +
   "      </entry>\n"+
   "      <entry>" +
   "         <string>three</string>\n" +
   "         <bigDecimal>3.0</bigDecimal>\n" +
   "      </entry>\n"+
   "      <entry>" +
   "         <string>four</string>\n" +
   "         <bigDecimal>4.0</bigDecimal>\n" +
   "      </entry>\n"+
   "   </table>\n"+
   "</primitiveMap>";  
   
   private static final String PRIMITIVE_VALUE_OVERRIDE_MAP =      
   "<primitiveValueOverrideMap>\n"+
   "   <map>\n"+   
   "      <entry>\n" +
   "         <string>one</string>\n" +
   "         <decimal>1.0</decimal>\n" +
   "      </entry>\n"+
   "      <entry>" +
   "         <string>two</string>\n" +
   "         <decimal>2.0</decimal>\n" +
   "      </entry>\n"+
   "      <entry>" +
   "         <string>three</string>\n" +
   "         <decimal>3.0</decimal>\n" +
   "      </entry>\n"+
   "      <entry>" +
   "         <string>four</string>\n" +
   "         <decimal>4.0</decimal>\n" +
   "      </entry>\n"+
   "   </map>\n"+
   "</primitiveValueOverrideMap>";
   
   private static final String PRIMITIVE_VALUE_KEY_OVERRIDE_MAP =      
   "<primitiveValueKeyOverrideMap>\n"+
   "   <map>\n"+   
   "      <item>\n" +
   "         <text>one</text>\n" +
   "         <decimal>1.0</decimal>\n" +
   "      </item>\n"+
   "      <item>" +
   "         <text>two</text>\n" +
   "         <decimal>2.0</decimal>\n" +
   "      </item>\n"+
   "      <item>" +
   "         <text>three</text>\n" +
   "         <decimal>3.0</decimal>\n" +
   "      </item>\n"+
   "      <item>" +
   "         <text>four</text>\n" +
   "         <decimal>4.0</decimal>\n" +
   "      </item>\n"+
   "   </map>\n"+
   "</primitiveValueKeyOverrideMap>"; 
   
   private static final String PRIMITIVE_INLINE_MAP =            
   "<primitiveInlineMap>\n"+ 
   "   <entity>\n" +
   "      <string>one</string>\n" +
   "      <bigDecimal>1.0</bigDecimal>\n" +
   "   </entity>\n"+
   "   <entity>" +
   "      <string>two</string>\n" +
   "      <bigDecimal>2.0</bigDecimal>\n" +
   "   </entity>\n"+
   "   <entity>" +
   "      <string>three</string>\n" +
   "      <bigDecimal>3.0</bigDecimal>\n" +
   "   </entity>\n"+
   "   <entity>" +
   "      <string>four</string>\n" +
   "      <bigDecimal>4.0</bigDecimal>\n" +
   "   </entity>\n"+
   "</primitiveInlineMap>";   
   
   
   @Root
   private static class EntryMap {      
      
      @ElementMap(key="key", attribute=true)
      private Map<String, MapEntry> map;
      
      public String getValue(String name) {
         return map.get(name).value;
      }
   }
   
   @Root
   private static class MapEntry {
      
      @Element
      private String name;
      
      @Element
      private String value;
      
      public MapEntry() {
         super();
      }
      
      public MapEntry(String name, String value) {
         this.name = name;
         this.value = value;
      }
   }
   
   @Root
   private static class StringMap {
      
      @ElementMap(key="letter", attribute=true, data=true)
      private Map<String, String> map;
      
      public String getValue(String name) {
         return map.get(name);
      }
   }
   
   @Root
   private static class ComplexMap {      
      
      @ElementMap
      private Map<CompositeKey, MapEntry> map;
      
      public ComplexMap() {
         this.map = new HashMap<CompositeKey, MapEntry>();
      }
      
      public String getValue(CompositeKey key) {
         return map.get(key).value;
      }
   }
   
   @Root
   private static class CompositeKey {
      
      @Element
      private String name;
      
      @Element
      private String address;
      
      public CompositeKey() {
         super();
      }
      
      public CompositeKey(String name, String address) {
         this.name = name;
         this.address = address;
      }
      
      public int hashCode() {
         return name.hashCode() + address.hashCode();
      }
      
      public boolean equals(Object item) {
         if(item instanceof CompositeKey) {
            CompositeKey other = (CompositeKey)item;
            
            return other.name.equals(name) && other.address.equals(address);
         }
         return false;
      }
   }
   
   @Root
   private static class PrimitiveMap {
      
      @ElementMap(name="table")
      private Map<String, BigDecimal> map;
      
      public PrimitiveMap() {
         this.map = new HashMap<String, BigDecimal>();         
      }
      
      public BigDecimal getValue(String name) {
         return map.get(name);
      }
   }
   
   public void testNullValue() throws Exception {
      Serializer serializer = new Persister();
      PrimitiveMap primitiveMap = new PrimitiveMap();
      
      primitiveMap.map.put("a", new BigDecimal(1));
      primitiveMap.map.put("b", new BigDecimal(2));
      primitiveMap.map.put("c", null);     
      primitiveMap.map.put(null, new BigDecimal(4));
      
      StringWriter out = new StringWriter();
      serializer.write(primitiveMap, out);
      
      primitiveMap = serializer.read(PrimitiveMap.class, out.toString());
      
      assertEquals(primitiveMap.map.get(null), new BigDecimal(4));
      assertEquals(primitiveMap.map.get("c"), null);
      assertEquals(primitiveMap.map.get("a"), new BigDecimal(1));
      assertEquals(primitiveMap.map.get("b"), new BigDecimal(2));
      
      validate(primitiveMap, serializer);           
      
      ComplexMap complexMap = new ComplexMap();
      
      complexMap.map.put(new CompositeKey("name.1", "address.1"), new MapEntry("1", "1"));
      complexMap.map.put(new CompositeKey("name.2", "address.2"), new MapEntry("2", "2"));
      complexMap.map.put(null, new MapEntry("3", "3"));
      complexMap.map.put(new CompositeKey("name.4", "address.4"), null);
      
      validate(complexMap, serializer);
   }
}

File: port/src/main/Xml/MapNullTest.cs
#region Using directives
using SimpleFramework.Xml.Core;
using SimpleFramework.Xml;
using System.Collections.Generic;
using System;
#endregion
namespace SimpleFramework.Xml.Core {
   public class MapNullTest : ValidationTestCase {
      private const String EMPTY_AS_NULL =
      "<complexMap>\r\n" +
      "   <map class='java.util.HashMap'>\r\n" +
      "      <entry>\r\n" +
      "         <mapEntry>\r\n" +
      "            <name>3</name>\r\n" +
      "            <value>3</value>\r\n" +
      "         </mapEntry>\r\n" +
      "      </entry>\r\n" +
      "      <entry>\r\n" +
      "         <compositeKey>\r\n" +
      "            <name>name.1</name>\r\n" +
      "            <address>address.1</address>\r\n" +
      "         </compositeKey>\r\n" +
      "         <mapEntry>\r\n" +
      "            <name>1</name>\r\n" +
      "            <value>1</value>\r\n" +
      "         </mapEntry>\r\n" +
      "      </entry>\r\n" +
      "      <entry>\r\n" +
      "         <compositeKey>\r\n" +
      "            <name>name.2</name>\r\n" +
      "            <address>address.2</address>\r\n" +
      "         </compositeKey>\r\n" +
      "         <mapEntry>\r\n" +
      "            <name>2</name>\r\n" +
      "            <value>2</value>\r\n" +
      "         </mapEntry>\r\n" +
      "      </entry>\r\n" +
      "      <entry>\r\n" +
      "         <compositeKey>\r\n" +
      "            <name>name.4</name>\r\n" +
      "            <address>address.4</address>\r\n" +
      "         </compositeKey>\r\n" +
      "      </entry>\r\n" +
      "   </map>\r\n" +
      "</complexMap>\r\n";
      private const String EMPTY_COMPOSITE_VALUE =
      "<complexMap>\r\n" +
      "   <map class='java.util.HashMap'>\r\n" +
      "      <entry>\r\n" +
      "         <compositeKey>\r\n" +
      "            <name>name.4</name>\r\n" +
      "            <address>address.4</address>\r\n" +
      "         </compositeKey>\r\n" +
      "      </entry>\r\n" +
      "   </map>\r\n" +
      "</complexMap>\r\n";
      private const String EMPTY_COMPOSITE_BLANK_VALUE =
      "<complexMap>\r\n" +
      "   <map class='java.util.HashMap'>\r\n" +
      "      <entry>\r\n" +
      "         <compositeKey>\r\n" +
      "            <name>name.4</name>\r\n" +
      "            <address>address.4</address>\r\n" +
      "         </compositeKey>\r\n" +
      "         <mapEntry/>\r\n" +
      "      </entry>\r\n" +
      "   </map>\r\n" +
      "</complexMap>\r\n";
      private const String EMPTY_COMPOSITE_KEY =
      "<complexMap>\r\n" +
      "   <map class='java.util.HashMap'>\r\n" +
      "      <entry>\r\n" +
      "         <mapEntry>\r\n" +
      "            <name>3</name>\r\n" +
      "            <value>3</value>\r\n" +
      "         </mapEntry>\r\n" +
      "      </entry>\r\n" +
      "   </map>\r\n" +
      "</complexMap>\r\n";
      private const String EMPTY_COMPOSITE_BLANK_KEY =
      "<complexMap>\r\n" +
      "   <map class='java.util.HashMap'>\r\n" +
      "      <entry>\r\n" +
      "         <compositeKey/>\r\n" +
      "         <mapEntry>\r\n" +
      "            <name>3</name>\r\n" +
      "            <value>3</value>\r\n" +
      "         </mapEntry>\r\n" +
      "      </entry>\r\n" +
      "   </map>\r\n" +
      "</complexMap>\r\n";
      private const String EMPTY_PRIMITIVE_VALUE =
      "<primitiveMap>\r\n" +
      "   <table class='java.util.HashMap'>\r\n" +
      "      <entry>\r\n" +
      "         <string>example</string>\r\n" +
      "      </entry>\r\n" +
      "   </table>\r\n" +
      "</primitiveMap>\r\n";
      private const String EMPTY_PRIMITIVE_BLANK_VALUE =
      "<primitiveMap>\r\n" +
      "   <table class='java.util.HashMap'>\r\n" +
      "      <entry>\r\n" +
      "         <string>example</string>\r\n" +
      "         <bigDecimal/>\r\n" +
      "      </entry>\r\n" +
      "   </table>\r\n" +
      "</primitiveMap>\r\n";
      private const String EMPTY_PRIMITIVE_KEY =
      "<primitiveMap>\r\n" +
      "   <table class='java.util.HashMap'>\r\n" +
      "      <entry>\r\n" +
      "         <bigDecimal>4</bigDecimal>\r\n" +
      "      </entry>\r\n" +
      "   </table>\r\n" +
      "</primitiveMap>\r\n";
      private const String EMPTY_PRIMITIVE_BLANK_KEY =
      "<primitiveMap>\r\n" +
      "   <table class='java.util.HashMap'>\r\n" +
      "      <entry>\r\n" +
      "         <string/>\r\n" +
      "         <bigDecimal>4</bigDecimal>\r\n" +
      "      </entry>\r\n" +
      "   </table>\r\n" +
      "</primitiveMap>\r\n";
      @Root
      private static class MapEntry {
         @Element
         private String name;
         @Element
         private String value;
         public MapEntry() {
            super();
         }
         public MapEntry(String name, String value) {
            this.name = name;
            this.value = value;
         }
         public bool Equals(Object other) {
            if(other instanceof MapEntry) {
               MapEntry entry = (MapEntry) other;
               if(entry.value.Equals(value)) {
                  return entry.name.Equals(name);
               }
            }
            return false;
         }
      }
      @Root
      private static class ComplexMap {
         @ElementMap
         private Map<CompositeKey, MapEntry> map;
         public ComplexMap() {
            this.map = new HashMap<CompositeKey, MapEntry>();
         }
         public String GetValue(CompositeKey key) {
            MapEntry entry = map.get(key);
            if(entry != null) {
               return entry.value;
            }
            return null;
         }
      }
      @Root
      private static class CompositeKey {
         @Element
         private String name;
         @Element
         private String address;
         public CompositeKey() {
            super();
         }
         public CompositeKey(String name, String address) {
            this.name = name;
            this.address = address;
         }
         public int HashCode() {
            return name.HashCode() + address.HashCode();
         }
         public bool Equals(Object item) {
            if(item instanceof CompositeKey) {
               CompositeKey other = (CompositeKey)item;
               return other.name.Equals(name) && other.address.Equals(address);
            }
            return false;
         }
      }
      @Root
      private static class PrimitiveMap {
         [ElementMap(Name="table")]
         private Map<String, BigDecimal> map;
         public PrimitiveMap() {
            this.map = new HashMap<String, BigDecimal>();
         }
         public BigDecimal GetValue(String name) {
            return map.get(name);
         }
      }
      public void TestEmptyCompositeValue() {
         Serializer serializer = new Persister();
         ComplexMap value = serializer.read(ComplexMap.class, EMPTY_COMPOSITE_VALUE);
         bool valid = serializer.validate(ComplexMap.class, EMPTY_COMPOSITE_VALUE);
         assertTrue(valid);
         validate(value, serializer);
      }
      public void TestEmptyCompositeBlankValue() {
         Serializer serializer = new Persister();
         ComplexMap value = serializer.read(ComplexMap.class, EMPTY_COMPOSITE_BLANK_VALUE);
         bool valid = serializer.validate(ComplexMap.class, EMPTY_COMPOSITE_BLANK_VALUE);
         assertTrue(valid);
         validate(value, serializer);
      }
      public void TestEmptyCompositeKey() {
         Serializer serializer = new Persister();
         ComplexMap value = serializer.read(ComplexMap.class, EMPTY_COMPOSITE_KEY);
         bool valid = serializer.validate(ComplexMap.class, EMPTY_COMPOSITE_KEY);
         assertTrue(valid);
         validate(value, serializer);
      }
      public void TestEmptyCompositeBlankKey() {
         Serializer serializer = new Persister();
         ComplexMap value = serializer.read(ComplexMap.class, EMPTY_COMPOSITE_BLANK_KEY);
         bool valid = serializer.validate(ComplexMap.class, EMPTY_COMPOSITE_BLANK_KEY);
         assertTrue(valid);
         validate(value, serializer);
      }
      public void TestEmptyPrimitiveValue() {
         Serializer serializer = new Persister();
         PrimitiveMap value = serializer.read(PrimitiveMap.class, EMPTY_PRIMITIVE_VALUE);
         bool valid = serializer.validate(PrimitiveMap.class, EMPTY_PRIMITIVE_VALUE);
         assertTrue(valid);
         validate(value, serializer);
      }
      public void TestEmptyPrimitiveBlankValue() {
         Serializer serializer = new Persister();
         PrimitiveMap value = serializer.read(PrimitiveMap.class, EMPTY_PRIMITIVE_BLANK_VALUE);
         bool valid = serializer.validate(PrimitiveMap.class, EMPTY_PRIMITIVE_BLANK_VALUE);
         assertTrue(valid);
         validate(value, serializer);
      }
      public void TestEmptyPrimitiveKey() {
         Serializer serializer = new Persister();
         PrimitiveMap value = serializer.read(PrimitiveMap.class, EMPTY_PRIMITIVE_KEY);
         bool valid = serializer.validate(PrimitiveMap.class, EMPTY_PRIMITIVE_KEY);
         assertTrue(valid);
         validate(value, serializer);
      }
      public void TestEmptyPrimitiveBlankKey() {
         Serializer serializer = new Persister();
         PrimitiveMap value = serializer.read(PrimitiveMap.class, EMPTY_PRIMITIVE_BLANK_KEY);
         bool valid = serializer.validate(PrimitiveMap.class, EMPTY_PRIMITIVE_BLANK_KEY);
         assertTrue(valid);
         validate(value, serializer);
      }
      public void TestNullValue() {
         Serializer serializer = new Persister();
         PrimitiveMap primitiveMap = new PrimitiveMap();
         primitiveMap.map.put("a", new BigDecimal(1));
         primitiveMap.map.put("b", new BigDecimal(2));
         primitiveMap.map.put("c", null);
         primitiveMap.map.put(null, new BigDecimal(4));
         StringWriter out = new StringWriter();
         serializer.write(primitiveMap, out);
         primitiveMap = serializer.read(PrimitiveMap.class, out.toString());
         assertEquals(primitiveMap.map.get(null), new BigDecimal(4));
         assertEquals(primitiveMap.map.get("c"), null);
         assertEquals(primitiveMap.map.get("a"), new BigDecimal(1));
         assertEquals(primitiveMap.map.get("b"), new BigDecimal(2));
         validate(primitiveMap, serializer);
         ComplexMap complexMap = new ComplexMap();
         complexMap.map.put(new CompositeKey("name.1", "address.1"), new MapEntry("1", "1"));
         complexMap.map.put(new CompositeKey("name.2", "address.2"), new MapEntry("2", "2"));
         complexMap.map.put(null, new MapEntry("3", "3"));
         complexMap.map.put(new CompositeKey("name.4", "address.4"), null);
         validate(complexMap, serializer);
         ComplexMap emptyNull = serializer.read(ComplexMap.class, EMPTY_AS_NULL);
         assertEquals(emptyNull.GetValue(new CompositeKey("name.1", "address.1")), "1");
         assertEquals(emptyNull.GetValue(new CompositeKey("name.2", "address.2")), "2");
         assertEquals(emptyNull.GetValue(null), "3");
         assertEquals(emptyNull.GetValue(new CompositeKey("name.4", "address.4")), null);
         validate(emptyNull, serializer);
      }
      // TODO test the null values and exceptions with the map
   }
}
